File: src/imagebuilder/flyingtux/services/x11.py
# SPDX-License-Identifier: AGPL-3.0-or-later
from .os_service import Base
from metux.util.fs import mkdir
from metux.util.log import warn
from os.path import dirname
from subprocess import Popen, PIPE, check_output
from os import environ, urandom, chmod
import re

"""
Access to X11 display.

2DO: add user-configured redirection to somewhere else, e.g.
a nested Xserver, so app has no access to X resources of other
applications.

X-NAMESPACE (see setup_namespace() below) confines the app to its own
namespace instead of getting the full, unrestricted display. Requires an
XLibre server with X-NAMESPACE support (merged in xserver-master as of
2026-07-03, PR #3103) started with "-namespace <config>" granting
management rights.

Falls back to full display access if X-NAMESPACE is unavailable (server
too old, not built with namespace support, or lacking management
privileges). A visible warning is printed on the console in this case.
"""
class X11(Base):
    permissions = {
        '__enabled__': { 'default': True },
        'shm':       { 'default': True, 'description': 'shared memory extension' },
        'namespace': { 'default': True, 'description': 'confine the app to its own X-NAMESPACE, if the display supports it and we are allowed to manage it (falls back to full access otherwise)' },
    }

    settings = {
        'x11-display': {
            'default':     '${ENV::DISPLAY}',
            'type':        'string',
            'description': 'number of the X11 display to attach to',
        },
        'x11-socketdir': {
            'default':     '/tmp/.X11-unix',
            'type':        'host-path',
            'description': 'directory of X11 sockets',
        },
        'x11-namespace-bin': {
            'default':     'xnamespace',
            'type':        'string',
            'description': 'path/name of the xnamespace CLI tool (github.com/X11Libre/go-x11proto) used to probe/manage the X-NAMESPACE extension',
        },
        'x11-namespace-caps': {
            'default':     'mouse,keyboard,shape,input',
            'type':        'string',
            'description': 'comma-separated X-NAMESPACE capabilities granted to the app namespace (never include admin/all here)',
        },
    }

    def compute(self):
        import re

        display = self.get_setting('x11-display')

        disp = re.search('^:([0-9]+)(\.[0-9]+|)$', display).group(1)
        sockname = self.get_path_setting('x11-socketdir')+'/X'+disp

        self.info("display: "+str(display))
        self.add_env('DISPLAY', display)
        self.bind_file_direct(sockname)

        if self.is_permitted('shm'):
            self.info("host shm: enabled")
            self.set_container_opt('ipc', 'host')
        else:
            self.info("host shm: disabled")

        if self.is_permitted('namespace'):
            self.setup_namespace(display)
        else:
            self.info("X-NAMESPACE isolation: disabled by config")

    def run_xnamespace(self, display, args):
        xnbin = self.get_setting('x11-namespace-bin')
        env = dict(environ)
        env['DISPLAY'] = display
        try:
            p = Popen([xnbin, '-s']+args, stdout=PIPE, stderr=PIPE, env=env)
        except OSError as e:
            return 127, '', str(e)
        stdout, stderr = p.communicate()
        return p.wait(), stdout.decode().strip(), stderr.decode().strip()

    """probe whether we're allowed to see/manage X-NAMESPACE on this display at all"""
    def probe_namespace(self, display):
        rc, out, err = self.run_xnamespace(display, ['version'])
        if rc != 0:
            self.info("X-NAMESPACE probe failed on "+display+": "+(err or "unknown error"))
            return False
        self.info("X-NAMESPACE present, version "+out)
        return True

    def setup_namespace(self, display):
        if not self.probe_namespace(display):
            warn("=" * 60)
            warn("X-NAMESPACE ISOLATION UNAVAILABLE")
            warn("App '"+self.get_app_name()+"' will have FULL access to the display")
            warn("No per-app isolation - all X11 resources are shared")
            warn("Start server with '-namespace <config>' to enable isolation")
            warn("=" * 60)
            return

        name = 'app-'+self.get_app_name()
        caps = [c for c in self.get_setting('x11-namespace-caps').split(',') if c]

        rc, out, err = self.run_xnamespace(display, ['create', name]+caps+['transient'])
        if rc != 0:
            warn("X-NAMESPACE create "+name+" failed (may already exist): "+(err or "unknown error"))
            warn("Falling back to full display access - no isolation")
            return

        token = urandom(16).hex()
        rc, out, err = self.run_xnamespace(display, ['addtoken', name, 'MIT-MAGIC-COOKIE-1', token])
        if rc != 0:
            warn("X-NAMESPACE addtoken for "+name+" failed: "+(err or "unknown error"))
            warn("Falling back to full display access - no isolation")
            return

        xauth_file = self.write_xauth(display, token)
        self.add_env('XAUTHORITY', '/etc/x11-app.xauth')
        self.bind_dir(xauth_file, '/etc/x11-app.xauth')
        self.info("X-NAMESPACE: confined to namespace '"+name+"' (caps: "+','.join(caps)+")")

    def write_xauth(self, display, token):
        fn = self.my_runner['TARGET::dynconf-dir']+'/xauth/'+self.get_app_name()+'.xauth'
        mkdir(dirname(fn))
        p = Popen(['xauth', '-f', fn, 'add', display, 'MIT-MAGIC-COOKIE-1', token], stdout=PIPE, stderr=PIPE)
        p.communicate()

        # xauth just wrote a FamilyLocal entry keyed to *our* hostname. The
        # container that reads this file has a different hostname, so a
        # FamilyLocal lookup from inside it never matches and the client
        # falls back to no auth data at all - silently defeating the whole
        # point of this per-app cookie. Rewrite the family as FamilyWild
        # (matches from any host) using the standard xauth nlist/nmerge trick.
        nlist = check_output(['xauth', '-f', fn, 'nlist', display])
        wild = re.sub(rb'^[0-9a-f]{4}', b'ffff', nlist)
        p2 = Popen(['xauth', '-f', fn, 'nmerge', '-'], stdin=PIPE)
        p2.communicate(wild)

        chmod(fn, 0o644)
        return fn

File: internal/services/x11.go
// SPDX-License-Identifier: AGPL-3.0-or-later

package services

import (
	"crypto/rand"
	"encoding/hex"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/metux/flyingtux/internal/fs"
	"github.com/metux/flyingtux/internal/spec"
	"github.com/metux/flyingtux/internal/util"
)

var displayRe = regexp.MustCompile(`^:([0-9]+)(\.[0-9]+|)$`)

// x11 grants access to the X11 display, optionally confining the app to its own
// X-NAMESPACE when the server supports it (best-effort, falls back to full
// access). Port of services/x11.py.
type x11 struct{ base }

func newX11(sp *spec.SpecObj, r Runner) Service {
	s := &x11{base{
		name: "X11", spec: sp, runner: r,
		permDflt: map[string]bool{"__enabled__": true, "shm": true, "namespace": true},
		setDflt: map[string]string{
			"x11-display":        "${ENV::DISPLAY}",
			"x11-socketdir":      "/tmp/.X11-unix",
			"x11-namespace-bin":  "xnamespace",
			"x11-namespace-caps": "mouse,keyboard,shape,input",
		},
	}}
	s.self = s
	s.init()
	return s
}

func (s *x11) compute() error {
	display := s.getSetting("x11-display")
	m := displayRe.FindStringSubmatch(display)
	if m == nil {
		return util.ConfigError("x11: cannot parse display %q", display)
	}
	disp := m[1]
	sockname := s.getSetting("x11-socketdir") + "/X" + disp

	s.info("display: " + display)
	s.addEnv("DISPLAY", display)
	s.bindFileDirect(sockname)

	if s.isPermitted("shm") {
		s.info("host shm: enabled")
		s.setContainerOpt(func(o *Opts) { o.IPC = "host" })
	} else {
		s.info("host shm: disabled")
	}

	if s.isPermitted("namespace") {
		s.setupNamespace(display)
	} else {
		s.info("X-NAMESPACE isolation: disabled by config")
	}
	return nil
}

func (s *x11) runXnamespace(display string, args []string) (int, string, string) {
	xnbin := s.getSetting("x11-namespace-bin")
	cmd := exec.Command(xnbin, append([]string{"-s"}, args...)...)
	cmd.Env = append(os.Environ(), "DISPLAY="+display)
	var out, errb strings.Builder
	cmd.Stdout = &out
	cmd.Stderr = &errb
	if err := cmd.Run(); err != nil {
		if ee, ok := err.(*exec.ExitError); ok {
			return ee.ExitCode(), strings.TrimSpace(out.String()), strings.TrimSpace(errb.String())
		}
		return 127, "", err.Error()
	}
	return 0, strings.TrimSpace(out.String()), strings.TrimSpace(errb.String())
}

func (s *x11) probeNamespace(display string) bool {
	rc, out, errs := s.runXnamespace(display, []string{"version"})
	if rc != 0 {
		if errs == "" {
			errs = "unknown error"
		}
		s.info("X-NAMESPACE unavailable on " + display + ": " + errs)
		return false
	}
	s.info("X-NAMESPACE present, version " + out)
	return true
}

func (s *x11) setupNamespace(display string) {
	if !s.probeNamespace(display) {
		return
	}

	name := "app-" + s.appName()
	var caps []string
	for _, c := range strings.Split(s.getSetting("x11-namespace-caps"), ",") {
		if c != "" {
			caps = append(caps, c)
		}
	}

	if rc, _, errs := s.runXnamespace(display, append(append([]string{"create", name}, caps...), "transient")); rc != 0 {
		if errs == "" {
			errs = "unknown error"
		}
		s.info("X-NAMESPACE create " + name + " failed (may already exist): " + errs)
	}

	tokenBuf := make([]byte, 16)
	if _, err := rand.Read(tokenBuf); err != nil {
		s.info("X-NAMESPACE: failed generating token: " + err.Error())
		return
	}
	token := hex.EncodeToString(tokenBuf)

	if rc, _, errs := s.runXnamespace(display, []string{"addtoken", name, "MIT-MAGIC-COOKIE-1", token}); rc != 0 {
		if errs == "" {
			errs = "unknown error"
		}
		s.info("X-NAMESPACE addtoken for " + name + " failed, falling back to full access: " + errs)
		return
	}

	xauthFile, err := s.writeXauth(display, token)
	if err != nil {
		s.info("X-NAMESPACE: writing xauth failed: " + err.Error())
		return
	}
	s.addEnv("XAUTHORITY", "/etc/x11-app.xauth")
	s.bindDir(xauthFile, "/etc/x11-app.xauth")
	s.info("X-NAMESPACE: confined to namespace '" + name + "' (caps: " + strings.Join(caps, ",") + ")")
}

func (s *x11) writeXauth(display, token string) (string, error) {
	fn := s.runner.Get("TARGET::dynconf-dir") + "/xauth/" + s.appName() + ".xauth"
	if err := fs.Mkdir(filepath.Dir(fn)); err != nil {
		return "", err
	}
	exec.Command("xauth", "-f", fn, "add", display, "MIT-MAGIC-COOKIE-1", token).Run()
	os.Chmod(fn, 0o644)
	return fn, nil
}
